### PR TITLE
Use thread local with BraveSearchServiceWorkerHolder

### PR DESCRIPTION
**Pull request by cdesouza-chromium**
A CL has been landed in `cr113` to remove `base::ThreadLocal` variants, and replace their use with C++'s `thread_local`.

In this change, `base::ThreadLocal` is being replaced with `thread_local` in `BraveSearchServiceWorkerHolder`.

    Chromium change:
    https://chromium.googlesource.com/chromium/src/+/4f39823862fe722c32bd72bdb0e126e952574b79
    
    commit 4f39823862fe722c32bd72bdb0e126e952574b79
    Author: Peter Kasting <pkasting@chromium.org>
    Date:   Thu Mar 16 01:34:24 2023 +0000
    
        Remove ThreadLocal{Pointer,Boolean}.
    
        These are now unused; `thread_local T*/bool` is generally a better
        option.
    
        Bug: 1416710

Resolves https://github.com/brave/brave-browser/issues/29137

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by petemill — COMMENTED** [review]
This looks kind of ok to me although now even more verbose to add a handler. (Is this how upstream now does it for every single handler they want to add?) However, this is not code I've really worked on. I think I just moved something around in this file once, so I'm tagging some other more knowledgable reviewers in the hope that they have a moment to review.

**Comment by cdesouza-chromium** [comment]
Thanks @petemill for the review. To your point, yes this PR is making things more in accordance with upstream, but it is also containing all the TLS data to its own class, which is what upstream is doing, but also good practice, to avoid mistakes, and have all the lifetimes, and procedures associated with them tied together.

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (sangwoo108) [inline components/brave_search/renderer/brave_search_service_worker_holder.cc:72]: Smart trick!

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (sangwoo108) [inline components/brave_search/renderer/brave_search_service_worker_holder.cc:95]: totally optional: I think upstream folks recommend push_back() when adding already created instance.

https://groups.google.com/a/chromium.org/g/chromium-dev/c/7mJypsYz6AA

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (sangwoo108) [inline components/brave_search/renderer/brave_search_service_worker_holder.cc:79]: nit: here too - |js_handler|

**Review by sangwoo108 — APPROVED** [review]
LGTM but you might want to wait for the second approval from others 😄

**Review by SergeyZhukovsky — APPROVED** [review]
++

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (cdesouza-chromium) [reply components/brave_search/renderer/brave_search_service_worker_holder.cc:79]: Answered in the other comment

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (cdesouza-chromium) [reply components/brave_search/renderer/brave_search_service_worker_holder.cc:78]: Unfortunately I don't think that would work, as `content::WorkerThread::AddObserver`/`RemoveObserver` are static, and `base::ScopedObservation` is designed to be used with observers that have a source.

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (cdesouza-chromium) [reply components/brave_search/renderer/brave_search_service_worker_holder.cc:72]: Cheers mate.

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (cdesouza-chromium) [reply components/brave_search/renderer/brave_search_service_worker_holder.cc:95]: Done.

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (cdesouza-chromium) [reply components/brave_search/renderer/brave_search_service_worker_holder.cc:105]: Fixed (sorry, I had misunderstood your request)

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (cdesouza-chromium) [reply components/brave_search/renderer/brave_search_service_worker_holder.cc:79]: Fixed

File: components/brave_search/renderer/brave_search_service_worker_holder.cc
Comment (sangwoo108) [reply components/brave_search/renderer/brave_search_service_worker_holder.cc:78]: Oh, yes. You're right. Thanks!

**Pull request merged**
